soc/arm/nordic_nrf/validate_base_addresses.c

```diff
@@ -1,284 +1,128 @@
-
-#if defined(DT_NORDIC_NRF_USBD_USBD_0_BASE_ADDRESS)
-CHECK_ADDRESS(DT_NORDIC_NRF_USBD_USBD_0_BASE_ADDRESS, NRF_USBD);
-#endif
-
-#if defined(DT_INST_0_NORDIC_NRF_VMC_BASE_ADDRESS)
-CHECK_ADDRESS(DT_INST_0_NORDIC_NRF_VMC_BASE_ADDRESS, NRF_VMC);
-#endif
-
-#if defined(DT_NORDIC_NRF_WATCHDOG_WDT_0_BASE_ADDRESS)
-CHECK_ADDRESS(DT_NORDIC_NRF_WATCHDOG_WDT_0_BASE_ADDRESS, NRF_WDT0);
+#if defined(CONFIG_SOC_NRF9160) && !defined(NRF_PDM0) && defined(NRF_PDM)
+#define NRF_PDM0 NRF_PDM
 #endif
 
-#if defined(DT_NORDIC_NRF_WATCHDOG_WDT_1_BASE_ADDRESS)
-CHECK_ADDRESS(DT_NORDIC_NRF_WATCHDOG_WDT_1_BASE_ADDRESS, NRF_WDT1);
-#endif
+/**
+ * Check that a devicetree node's "reg" base address matches the
+ * correct value from the MDK.
+ *
+ * This check will only happen if the node is enabled. Disabled or
+ * nonexistent nodes are not checked.
+ *
+ * This allows the same file to work with multiple SoCs and devicetree
+ * configurations.
+ *
+ * @param lbl lowercase-and-underscores devicetree node label to check
+ * @param mdk_addr expected address from the Nordic MDK.
+ */
+#define CHECK_DT_REG(lbl, mdk_addr)					\
+	BUILD_ASSERT(							\
+		UTIL_OR(UTIL_NOT(DT_HAS_NODE(DT_NODELABEL(lbl))),	\
+			(DT_REG_ADDR(DT_NODELABEL(lbl)) ==		\
+			 (u32_t)(mdk_addr))))
+
+/**
+ * If a node label 'lbl' might have different addresses depending on
+ * its compatible 'compat', you can use this macro to pick the right
+ * one.
+ *
+ * @param lbl lowercase-and-underscores devicetree node label to check
+ * @param compat lowercase-and-underscores compatible to check
+ * @param addr_if_match MDK address to return if "lbl" has compatible "compat"
+ * @param addr_if_no_match MDK address to return otherwise
+ */
+#define NODE_ADDRESS(lbl, compat, addr_if_match, addr_if_no_match)	\
+	COND_CODE_1(DT_NODE_HAS_COMPAT(DT_NODELABEL(lbl), compat),	\
+		    (addr_if_match), (addr_if_no_match))
+
+#define ADC_ADDRESS NODE_ADDRESS(adc, nordic_nrf_adc, NRF_ADC, NRF_SAADC)
+
+#define SPI(num) spi##num
+
+#define SPI_ADDRESS(num) \
+	NODE_ADDRESS(SPI(num), nordic_nrf_spi, NRF_SPI##num,		\
+		NODE_ADDRESS(SPI(num), nordic_nrf_spis, NRF_SPIS##num,	\
+			     NRF_SPIM##num))
+
+#define I2C(num) i2c##num
+
+#define TWI_ADDRESS(num) \
+	NODE_ADDRESS(I2C(num), nordic_nrf_twi, NRF_TWI##num,		\
+		NODE_ADDRESS(I2C(num), nordic_nrf_twim, NRF_TWIM##num,	\
+			     NRF_TWIS##num))
+
+#define UART_ADDRESS(num) \
+	NODE_ADDRESS(uart##num, nordic_nrf_uart, NRF_UART##num, NRF_UARTE##num)
+
+CHECK_DT_REG(adc, ADC_ADDRESS);
+CHECK_DT_REG(clock, NRF_CLOCK);
+CHECK_DT_REG(cryptocell, NRF_CRYPTOCELL);
+CHECK_DT_REG(dppic, NRF_DPPIC);
+CHECK_DT_REG(egu0, NRF_EGU0);
+CHECK_DT_REG(egu1, NRF_EGU1);
+CHECK_DT_REG(egu2, NRF_EGU2);
+CHECK_DT_REG(egu3, NRF_EGU3);
+CHECK_DT_REG(egu4, NRF_EGU4);
+CHECK_DT_REG(egu5, NRF_EGU5);
+CHECK_DT_REG(ficr, NRF_FICR);
+CHECK_DT_REG(flash_controller, NRF_NVMC);
+CHECK_DT_REG(gpio0, NRF_P0);
+CHECK_DT_REG(gpio1, NRF_P1);
+CHECK_DT_REG(gpiote, NRF_GPIOTE);
+CHECK_DT_REG(i2s0, NRF_I2S);
```

Comment:
> It should be `NRF_I2S0` and corresponding translation should be added.

Is there a rule that determines when these should be added or not? E.g. there's no translation for NRF_CLOCK0 or NRF_QDEC0. Why is I2S different?

Reply:
That's because for I2S we get (from MDK) NRF_I2S0 symbol defined for nRF5340 and NRF_I2S for others. We need to handle both in the same driver (nrfx_i2s.c), so we translate NRF_I2S to NRF_I2S0 in nrf_i2s.h. Same story with NRF_PDM[0] nad NRF_WDT[0]. NRF_CLOCK and NRF_QDEC are always defined without indexes.

Reply:
Ah, I missed the 5340 case. Thanks.